pkg/agent/controller/bgp/controller.go

```diff
@@ -0,0 +1,950 @@
+			return err
+		}
+		c.bgpPolicyState.routes.Insert(route)
+	}
+	for route := range routesToWithdraw {
+		if err := bgpServer.WithdrawRoutes(c.ctx, []bgp.Route{route}); err != nil {
+			return err
+		}
+		c.bgpPolicyState.routes.Delete(route)
+	}
+
+	return nil
+}
+
+func hashNodeNameToIP(s string) string {
+	h := fnv.New32a() // Create a new FNV hash
+	h.Write([]byte(s))
+	hashValue := h.Sum32() // Get the 32-bit hash
+
+	// Convert the hash to a 4-byte slice
+	ip := make(net.IP, 4)
+	ip[0] = byte(hashValue >> 24)
+	ip[1] = byte(hashValue >> 16)
+	ip[2] = byte(hashValue >> 8)
+	ip[3] = byte(hashValue)
+
+	return ip.String()
+}
+
+func (c *Controller) getRouterID() (string, error) {
+	// According to RFC 4271:
+	// BGP Identifier:
+	//   This 4-octet unsigned integer indicates the BGP Identifier of
+	//   the sender.  A given BGP speaker sets the value of its BGP
+	//   Identifier to an IP address that is assigned to that BGP
+	//   speaker.  The value of the BGP Identifier is determined upon
+	//   startup and is the same for every local interface and BGP peer.
+	//
+	// In goBGP, only an IPv4 address can be used as the BGP Identifier (BGP router ID).
+	// For IPv4-only or dual-stack Kubernetes clusters, the Node's IPv4 address is used as the BGP router ID, ensuring
+	// uniqueness.
+	// For IPv6-only Kubernetes clusters without a Node IPv4 address, the router ID could be specified in the Node
+	// annotation `node.antrea.io/bgp-router-id`. If the annotation is not present, an IPv4 address will be generated by
+	// hashing the Node name and updated to the Node annotation `node.antrea.io/bgp-router-id`.
+
+	if c.enabledIPv4 {
+		return c.nodeIPv4Addr, nil
+	}
+
+	nodeObj, err := c.nodeLister.Get(c.nodeName)
+	if err != nil {
+		return "", fmt.Errorf("failed to get Node object: %w", err)
+	}
+
+	var exists bool
+	var routerID string
+	routerID, exists = nodeObj.GetAnnotations()[types.NodeBGPRouterIDAnnotationKey]
+	if !exists {
+		routerID = hashNodeNameToIP(c.nodeName)
+		patch, _ := json.Marshal(map[string]interface{}{
+			"metadata": map[string]interface{}{
+				"annotations": map[string]string{
+					types.NodeBGPRouterIDAnnotationKey: routerID,
+				},
+			},
+		})
+		if _, err := c.k8sClient.CoreV1().Nodes().Patch(context.TODO(), c.nodeName, apitypes.MergePatchType, patch, metav1.PatchOptions{}, "status"); err != nil {
+			return "", fmt.Errorf("failed to patch BGP router ID to Node annotation %s: %w", types.NodeBGPRouterIDAnnotationKey, err)
+		}
+	} else if !utilnet.IsIPv4String(routerID) {
+		return "", fmt.Errorf("BGP router ID should be an IPv4 address string")
+	}
+	return routerID, nil
+}
+
+func (c *Controller) getRoutes(advertisements v1alpha1.Advertisements) (sets.Set[bgp.Route], error) {
+	allRoutes := sets.New[bgp.Route]()
+
+	if advertisements.Service != nil {
+		c.addServiceRoutes(advertisements.Service, allRoutes)
+	}
+	if c.egressEnabled && advertisements.Egress != nil {
```

Comment:
I didn't see there is any validation about Egress in advertisement when the egress is not enabled, maybe add a log to highlight the Egress in advertisements is ignored when Egress is not enabled.
```
if !c.egressEnabled && advertisements.Egress != nil {
   Klog.InfoS(...)
}
```

Reply:
We cannot stop users from advertising Egress IPs even when Egress is disabled. IMO, that's what the users should know by themselves.

Reply:
If Egress is not enabled, none of the Egress resources won't work, regardless of whether it's advertised or not. The check here is for performance consideration: no need to start an informer for Egress if it's not enabled.